### PR TITLE
[SPARK-23099][SS] Migrate foreach sink to DataSourceV2

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Migrate foreach sink to DataSourceV2.

Since the previous attempt at this PR #20552, we've changed and strictly defined the lifecycle of writer components. This means we no longer need the complicated lifecycle shim from that PR; it just naturally works.

## How was this patch tested?

existing tests

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala [hunk 1]
```diff
@@ -17,52 +17,81 @@
...
 
-/**
- * A [[Sink]] that forwards all data into [[ForeachWriter]] according to the contract defined by
- * [[ForeachWriter]].
- *
- * @param writer The [[ForeachWriter]] to process all data.
- * @tparam T The expected type of the sink.
- */
-class ForeachSink[T : Encoder](writer: ForeachWriter[T]) extends Sink with Serializable {
+case class ForeachWriterProvider[T: Encoder](writer: ForeachWriter[T]) extends StreamWriteSupport {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: Rename the file accordingly. and Add docs.

Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: Actually, why not make it extend DataSourceV2 for consistency sake? Then it is easier to find all data sources in code by looking at who extends DataSourceV2

Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ForeachSink.scala]: This is a DataSourceV2 - that interface extends it

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 2]
```diff
@@ -131,7 +131,7 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
         .foreach(new TestForeachWriter() {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: Maybe rename this to `ForeachWriterSuite`?

Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: And move this to streaming.sources package similar ConsoleWriterSuite

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala [hunk 3]
```diff
@@ -141,7 +141,7 @@ class ForeachSinkSuite extends StreamTest with SharedSQLContext with BeforeAndAf
         query.processAllAvailable()
       }
       assert(e.getCause.isInstanceOf[SparkException])
-      assert(e.getCause.getCause.getMessage === "error")
+      assert(e.getCause.getCause.getCause.getMessage === "ForeachSinkSuite error")
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: why 3 levels? Can you paste the levels?

**Comment by tdas** [comment]
LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala (hunk 3 above)
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/ForeachSinkSuite.scala]: [info]   org.apache.spark.sql.streaming.StreamingQueryException: Query [id = c80c8860-d4f5-47c6-9a2b-33b5172e1735, runId = 81acd408-9028-41ee-9349-866ae2d67615] terminated with exception: Writing job aborted.
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:295)

[info]   Cause: org.apache.spark.SparkException: Writing job aborted.
[info]   at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec.doExecute(WriteToDataSourceV2.scala:117)

[info]   Cause: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 1, localhost, executor driver): java.lang.RuntimeException: ForeachSinkSuite error
[info] 	at org.apache.spark.sql.execution.streaming.sources.ForeachWriterSuite$$anonfun$5$$anonfun$apply$mcV$sp$4$$anon$1.process(ForeachWriterSuite.scala:135)

[info]   Cause: java.lang.RuntimeException: ForeachSinkSuite error
[info]   at org.apache.spark.sql.execution.streaming.sources.ForeachWriterSuite$$anonfun$5$$anonfun$apply$mcV$sp$4$$anon$1.process(ForeachWriterSuite.scala:135)

**Pull request closed without merging**